### ISSUE TITLE
libffi: configuration doesn't take libffi.pc into account, other problems with external libraries

### ISSUE THREAD

**Issue by mojca** [MoarVM/MoarVM#320]
I'm using OS X (10.7, x86_64) and am trying to make a package for MoarVM. It builds fine if I use all the included 3rdparty libraries, but if I fetch the tarball from GitHub, it fails to find `libffi`.

The list of files that `libffi` installs is indeed weird (and most likely wrong, I'll try to get that fixed):
```
  /opt/local/lib/libffi-3.2.1/include/ffi.h
  /opt/local/lib/libffi-3.2.1/include/ffitarget.h
```

but the fact is that `pkg-config` works:

```
> pkg-config --cflags libffi
-I/opt/local/lib/libffi-3.2.1/include
```

and I would find it more intuitive if MoarVM's configuration would use pkgconfig's hint about required flags.

I will try to report other problems separately, but here's a quick glimpse at another example. I wasn't able to use the built-in `libtommath` because MoarVM would then install the same files as already provided by `libtommath` and would thus conflict with that package.

But if I use `--has-libtommath`, then I run into:
```
compiling src/core/callsite.o
In file included from src/core/callsite.c:1:
In file included from src/moar.h:122:
In file included from src/6model/reprs.h:22:
src/6model/reprs/P6bigint.h:1:10: fatal error: 'tommath.h' file not found
```

Here are some files installed by `libtomath` package in my package manager:
```
/opt/local/include/libtommath/tommath.h
/opt/local/include/libtommath/tommath_class.h
/opt/local/include/libtommath/tommath_superclass.h
```

Apparently one level of `"libtommath"` is missing in the include statement. But sadly that library doesn't come with its own `.pc` file to be able to find it with `pkg-config`.

**Comment by cygx** [comment]
Just some general comments:

>  if I fetch the tarball from GitHub

Don't. Either use the tarballs from [moarvm.com](http://moarvm.com/), or clone the git repository. The tarballs auto-generated by github don't package the submodules.

As to `libtommath`, you probably should go with the packaged version (or the upstream `libtom` git repository). The distro packages might be outdated.

**Comment by mojca** [comment]
Yes, I noticed that tarballs from GitHub were problematic and switched, but I didn't know that when I started. In any case, even if the sources come from the official tarball, the issue about `pkg-config` is still true.

Regarding `libtommath`: thanks a lot for the warning. I checked and our package manager provides the latest release (we could probably switch to a more recent one as long as someone tells us which one to take), but yes, the released version is very old. The problem is that I'm unable to use the built-in version as long as it installs the files to the same location as `libtommath` does as that would make the packages conflict with each other, see #321.

Is there any other package where I should avoid using the released version?

**Comment by flussence** [comment]
In the meantime, here's a patch that fixes this:

    diff --git a/Configure.pl b/Configure.pl
    index 0ef15b6c96d4..6cd7d023895e 100755
    --- a/Configure.pl
    +++ b/Configure.pl
    @@ -230,6 +230,7 @@ else {
     
     if ($args{'has-libffi'}) {
         $config{nativecall_backend} = 'libffi';
    +    chomp($config{cincludes} .= ' ' . `pkg-config --cflags-only-I libffi`);
         unshift @{$config{usrlibs}}, 'ffi';
         push @{$config{defs}}, 'HAVE_LIBFFI';
     }

**Comment by flussence** [comment]
Any progress? It's been close to half a year.

**Comment by mojca** [comment]
Do you expect feedback from me (who reported the issue) or from developers?

**Comment by flussence** [comment]
They're the ones who have the commit access and the responsibility so...

I wasn't expecting a fix given how long most bugs here go ignored, but even a "we won't fix this build failure because..." would be *some* closure.

To be honest, I'm just tired of working around Debianisms for half a year.

**Comment by tomboy-64** [comment]
Please consider giving [PR 363](https://github.com/MoarVM/MoarVM/pull/363) a go. I ran into the exact same issue on Gentoo and decided to patch it.

**Comment by flussence** [comment]
This looks fixed, I can now get all the way into a REPL and invoke `rand(3)` using a `/usr/bin/moar` linked against libffi.

**Comment by mojca** [comment]
The only annoying thing from the fix in #363 is that:
* I have to manually specify `--pkgconfig=${prefix}/bin/pkg-config`; I would expect from the build system to be clever enough to find `pkg-config` on its own even if it is not in `/usr/bin/pkg-config`.
* I end up with `libs: -lffi -L/opt/local/lib -lffi  -latomic_ops -L/opt/local/lib -lffi  -luv -L/opt/local/lib -lffi  -lpthread -L/opt/local/lib -lffi`, so basically I get one copy of `-lffi` for every other library that's included. This looks like a bug.
